lib/OptimizationManopt/test/runtests.jl

```diff
@@ -134,4 +217,27 @@ end
     @time sol = Optimization.solve(prob, opt)
 
     @test sol.u≈q atol=1e-2
+
+    function closed_form_solution!(M::SymmetricPositiveDefinite, q, L, U, p, X)
+        # extract p^1/2 and p^{-1/2}
+        (p_sqrt_inv, p_sqrt) = Manifolds.spd_sqrt_and_sqrt_inv(p)
+        # Compute D & Q
+        e2 = eigen(p_sqrt_inv * X * p_sqrt_inv) # decompose Sk  = QDQ'
+        D = Diagonal(1.0 .* (e2.values .< 0))
+        Q = e2.vectors
+        #println(p)
+        Uprime = Q' * p_sqrt_inv * U * p_sqrt_inv * Q
+        Lprime = Q' * p_sqrt_inv * L * p_sqrt_inv * Q
+        P = cholesky(Hermitian(Uprime - Lprime))
+        z = P.U' * D * P.U + Lprime
+        copyto!(M, q, p_sqrt * Q * z * Q' * p_sqrt)
+        return q
+    end
+    N = m
+    U = mean(data2)
+    L = inv(sum(1/N * inv(matrix) for matrix in data2))
+
+    optf = OptimizationFunction(f, Optimization.AutoZygote())
+    prob = OptimizationProblem(optf, U; manifold = M, maxiters = 1000)
+    @time sol = Optimization.solve(prob, opt, sub_problem = (M, q, p, X) -> closed_form_solution!(M, q, L, U, p, X))
```

Comment:
This is also a bit surprising that this actually takes as long as the naive Gradient Descent approach above, might be something on my end but was wondering if either of you had any suspicions what might be wrong here, I am going to profile etc to see if it's a code issue

Reply:
I yet again struggle a bit with what this is doing.

1) Why does f have an option parameter `p` that is not used?
2) this seams to compute the Riemannian Center of mass (also called Karcher mean, so the generalisation of a mean to data on a manifold, here SPDs). This does not have a closed form solution. So I do not understand what this is doing.
3) I am confused by this last line that has two assignments. Am I missing something? What is this doing?

Reply:
4. `optf` seems to do some “Euclidean Gradient computation” but I can not see how the solver magically know these are Euclidean and not Riemannian?

Reply:
>Why does f have an option parameter p that is not used?

"p: the constant parameters used for defining the problem. Defaults to NullParameters." https://docs.sciml.ai/Optimization/stable/API/optimization_problem/

>this seams to compute the Riemannian Center of mass (also called Karcher mean, so the generalisation of a mean to data on a manifold, here SPDs). This does not have a closed form solution. So I do not understand what this is doing.

Apologies for the confusion. This test is intended to use the FW solver not GD here, hence the sub_problem with closed form.

>optf seems to do some “Euclidean Gradient computation” but I can not see how the solver magically know these are Euclidean and not Riemannian?

https://github.com/SciML/Optimization.jl/blob/f2d3e6358564cf5302193e4eec371b1335ddc72a/lib/OptimizationManopt/src/OptimizationManopt.jl#L202-L212 converts it from Euclidean to riemannian, that was part of the previous PR when we got the first order methods working. When passed to the solver it's always the riemannian gradient https://github.com/SciML/Optimization.jl/blob/master/lib/OptimizationManopt/src/OptimizationManopt.jl#L286 https://github.com/SciML/Optimization.jl/blob/master/lib/OptimizationManopt/src/OptimizationManopt.jl#L294

Reply:
Ah, but Frank Wolfe solves something different than GD here. The sub problem (you took from https://juliamanifolds.github.io/ManoptExamples.jl/stable/examples/Difference-of-Convex-Frank-Wolfe/ probably) is a _constraint_ Riemannian Center of Mass.
Compared to what I learned (when writing the linked example) is that FW is not really fast on this constrained problem.

Reply:
> When passed to the solver it's always the riemannian gradient

My question was more what it is _before_. So you assume that the user always calls the Optimization interface with the Euclidean?

Reply:
Concerning the last PR – hm I did not follow that one up totally, usually your PRs here are too fast for me as a developer that usually mainly works a bit on the weekends and in the evenings on code, since during the week I usually have a lot of teaching and supervision to do. So I can not claim of having understood much (if anything) of the last PR).

Reply:
>Compared to what I learned (when writing the linked example) is that FW is not really fast on this constrained problem.

Ah that's surprising since I expected with the closed form subproblem it should be faster as the original Sra and Weber mentions.

>My question was more what it is before. So you assume that the user always calls the Optimization interface with the Euclidean?

Yes, and when not passed the AD backend generates the Euclidean grad/hess that's then converted to the riemannian

Reply:
>Ah that's surprising since I expected with the closed form subproblem it should be faster as the original Sra and Weber mentions.

But that seems to have some NaNs and Infs showing up somewhere as well so couldn't verify this yet

```
ERROR: ArgumentError: matrix contains Infs or NaNs
Stacktrace:
  [1] chkuplofinite(A::Matrix{Float64}, uplo::Char)
    @ LinearAlgebra.LAPACK ~/.julia/juliaup/julia-1.10.0+0.x64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/lapack.jl:98
  [2] syevr!(jobz::Char, range::Char, uplo::Char, A::Matrix{…}, vl::Float64, vu::Float64, il::Int64, iu::Int64, abstol::Float64)
    @ LinearAlgebra.LAPACK ~/.julia/juliaup/julia-1.10.0+0.x64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/lapack.jl:5222
  [3] eigen!
    @ ~/.julia/juliaup/julia-1.10.0+0.x64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/symmetriceigen.jl:8 [inlined]
  [4] eigen(A::Symmetric{Float64, Matrix{Float64}}; sortby::Nothing)
    @ LinearAlgebra ~/.julia/juliaup/julia-1.10.0+0.x64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/symmetriceigen.jl:13
  [5] eigen
    @ ~/.julia/juliaup/julia-1.10.0+0.x64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/symmetriceigen.jl:11 [inlined]
  [6] log!(::SymmetricPositiveDefinite{…}, X::Matrix{…}, p::Matrix{…}, q::Matrix{…})
    @ Manifolds ~/.julia/packages/Manifolds/QwbWu/src/manifolds/SymmetricPositiveDefiniteAffineInvariant.jl:353
  [7] log
    @ ~/.julia/packages/ManifoldsBase/ggYC2/src/exp_log_geo.jl:150 [inlined]
  [8] log(::ManifoldsBase.EmptyTrait, M::SymmetricPositiveDefinite{…}, p::Matrix{…}, q::Matrix{…})
    @ ManifoldsBase ~/.julia/packages/ManifoldsBase/ggYC2/src/nested_trait.jl:346
  [9] log(::ManifoldsBase.EmptyTrait, M::SymmetricPositiveDefinite{…}, p::Matrix{…}, q::Matrix{…}) (repeats 3 times)
    @ ManifoldsBase ~/.julia/packages/ManifoldsBase/ggYC2/src/nested_trait.jl:313 [inlined]
 [10] log(::ManifoldsBase.EmptyTrait, M::SymmetricPositiveDefinite{…}, p::Matrix{…}, q::Matrix{…})
    @ ManifoldsBase ~/.julia/packages/ManifoldsBase/ggYC2/src/nested_trait.jl:306 [inlined]
 [11] shortest_geodesic
    @ ~/.julia/packages/ManifoldsBase/ggYC2/src/exp_log_geo.jl:183 [inlined]
 [12] adjoint_Jacobi_field!(M::SymmetricPositiveDefinite{…}, Y::Matrix{…}, p::Matrix{…}, q::Matrix{…}, t::Float64, X::Matrix{…}, β::typeof(ManifoldDiff.βdifferential_log_argument))
    @ ManifoldDiff ~/.julia/packages/ManifoldDiff/TFtNd/src/Jacobi_fields.jl:149
 [13] adjoint_Jacobi_field
    @ ~/.julia/packages/ManifoldDiff/TFtNd/src/Jacobi_fields.jl:141 [inlined]
 [14] adjoint_differential_log_argument
    @ ~/.julia/packages/ManifoldDiff/TFtNd/src/adjoint_differentials.jl:135 [inlined]
 [15] FrankWolfeGradient
    @ ~/.julia/dev/Manopt/src/plans/frank_wolfe_plan.jl:46 [inlined]
 [16] get_gradient
    @ ~/.julia/dev/Manopt/src/plans/gradient_plan.jl:171 [inlined]
 [17] get_gradient
    @ ~/.julia/dev/Manopt/src/plans/gradient_plan.jl:142 [inlined]
 [18] ArmijoLinesearch
    @ ~/.julia/dev/Manopt/src/plans/stepsize.jl:292 [inlined]
 [19] #_get_stepsize#152
    @ ~/.julia/dev/Manopt/src/plans/stepsize.jl:1240 [inlined]
 [20] _get_stepsize
    @ ~/.julia/dev/Manopt/src/plans/stepsize.jl:1233 [inlined]
 [21] #get_stepsize#150
    @ ~/.julia/dev/Manopt/src/plans/stepsize.jl:1222 [inlined]
 [22] get_stepsize
    @ ~/.julia/dev/Manopt/src/plans/stepsize.jl:1219 [inlined]
 [23] (::IdentityUpdateRule)(mp::DefaultManoptProblem{…}, s::GradientDescentState{…}, i::Int64)
    @ Manopt ~/.julia/dev/Manopt/src/solvers/gradient_descent.jl:80
 [24] step_solver!(p::AbstractManoptProblem, s::GradientDescentState, i::Any)
    @ Manopt ~/.julia/dev/Manopt/src/solvers/gradient_descent.jl:279 [inlined]
 [25] solve!(p::DefaultManoptProblem{…}, s::GradientDescentState{…})
    @ Manopt ~/.julia/dev/Manopt/src/solvers/solver.jl:137
 [26] step_solver!(amp::DefaultManoptProblem{…}, fws::FrankWolfeState{…}, i::Int64)
    @ Manopt ~/.julia/dev/Manopt/src/solvers/FrankWolfe.jl:321
 [27] step_solver!
    @ ~/.julia/dev/Manopt/src/solvers/solver.jl:150 [inlined]
 [28] solve!(p::DefaultManoptProblem{…}, s::Manopt.ReturnSolverState{…})
    @ Manopt ~/.julia/dev/Manopt/src/solvers/solver.jl:137
 [29] Frank_Wolfe_method!(M::SymmetricPositiveDefinite{…}, mgo::ManifoldGradientObjective{…}, p::Matrix{…}; initial_vector::Matrix{…}, evaluation::InplaceEvaluation, objective_type::Symbol, retraction_method::ExponentialRetraction, stepsize::DecreasingStepsize, stopping_criterion::StopWhenAny{…}, sub_cost::FrankWolfeCost{…}, sub_grad::FrankWolfeGradient{…}, sub_kwargs::@NamedTuple{}, sub_objective::ManifoldGradientObjective{…}, sub_problem::DefaultManoptProblem{…}, sub_stopping_criterion::StopWhenAny{…}, sub_state::GradientDescentState{…}, kwargs::@Kwargs{…})
    @ Manopt ~/.julia/dev/Manopt/src/solvers/FrankWolfe.jl:305
 [30] #Frank_Wolfe_method#708
    @ ~/.julia/dev/Manopt/src/solvers/FrankWolfe.jl:224 [inlined]
 [31] Frank_Wolfe_method
    @ ~/.julia/dev/Manopt/src/solvers/FrankWolfe.jl:220 [inlined]
 [32] #Frank_Wolfe_method#705
    @ ~/.julia/dev/Manopt/src/solvers/FrankWolfe.jl:202 [inlined]
 [33] Frank_Wolfe_method
    @ ~/.julia/dev/Manopt/src/solvers/FrankWolfe.jl:193 [inlined]
```

Reply:
Hm, sorry, I do not follow you.
Yes we can set up an subsolver in FW in Manopt.jl. Usually that should be a bit slower than the closed form one, but not too mach.
FW itself is not really fast by itself, so it should not make so much of a difference.

If the linesearch / geodesic hits you with NaNs, that is usually due to not providing a tangent vector (on SPDs). Much more I can not say without running that myself.

Reply:
Have you checked if Zygote actually returns correct gradients here? Zygote and FiniteDifferences.jl return completely different results but I haven't investigated it thoroughly.

Reply:
Most of Manifolds.jl wasn't designed with AD-friendliness in mind so ADing through things like `distance` requires careful checking for correctness.

Reply:
Ah interesting, I didn't think about that. I'll try out other combinations thanks!

Reply:
I tried making Manifolds.jl more AD friendly at some point but it was a very tedious whack-a-mole with lots of performance issues and conflicting viewpoints on what it should do so I gave up. When AD works, I'm happy, when it doesn't, I just try to derive things by hand.

Reply:
yeah I can verify it works with forwarddiff and finitediff, zygote seems to be the issue here